tools/squidclient.1
.if !'po4a'hide' .TH squidclient 1
.
.SH NAME
squidclient \- A simple HTTP web client tool
.
.SH SYNOPSIS
.if !'po4a'hide' .B squidclient
.if !'po4a'hide' .B "[ \-aknNrsv ] [ \-A"
string
.if !'po4a'hide' .B "] [ \-g"
count
.if !'po4a'hide' .B "] [ \-h"
remote host
.if !'po4a'hide' .B "] [ \-H '"
string
.if !'po4a'hide' .B "' ] [ \-i"
IMS
.if !'po4a'hide' .B "] [ \-I"
ping interval
.if !'po4a'hide' .B "] [ \-j '"
Host header
.if !'po4a'hide' .B "' ] [ \-l"
local host
.if !'po4a'hide' .B "] [ \-m"
method
.if !'po4a'hide' .B "] [ \-p"
port
.if !'po4a'hide' .B "] [ \-P"
file
.if !'po4a'hide' .B "] [ \-t"
count
.if !'po4a'hide' .B "] [ \-T"
timeout
.if !'po4a'hide' .B "] [ \-u"
user
.if !'po4a'hide' .B "] [ \-U"
user
.if !'po4a'hide' .B "] [ \-V"
version
.if !'po4a'hide' .B "] [ \-w"
password
.if !'po4a'hide' .B "] [ \-W"
password
.if !'po4a'hide' .B "] "
url
.
.SH DESCRIPTION
.B squidclient
is a tool providing a command line interface for retrieving URLs.
Designed for testing any HTTP 0.9, 1.0, or 1.1 web server or proxy.
This tool can be combined with scripts to perform any basic HTTP operation.
Some additional features for access to the 
.B squid
proxy object cache and management information are provided.
.
.SH OPTIONS
.if !'po4a'hide' .TP 12
.if !'po4a'hide' .B "\-a"
Do NOT include Accept: header.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-A 'string'"
Send 
.B string
as User-Agent: header. To omit the header completely set string to empty ('').
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-g count"
Ping mode, perform
.I count
iterations (0 to loop until interrupted).
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-h host"
Retrieve URL from cache on hostname.  Default is
.B localhost
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-H 'string'"
Extra headers to send. Use
.B '\\n'
for new lines.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-i time"
If\-Modified\-Since time (in Epoch seconds).
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-I interval"
Ping interval in seconds (default 1 second).
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-j hosthdr"
Host header content
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-k"
Keep the connection active. Default is to do only one request then close.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-l host"
Specify a local IP address to bind to.  Default is none.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-m method"
Request method, default is
.I GET.
Squid also supports a non-standard method called
.I PURGE.
You can use that to purge a specific URL from the cache.
You need to have
.I purge
access setup in
.B squid.conf
similar to
.I manager
 access. Here is an example:
.if !'po4a'hide' .nf
.if !'po4a'hide'       acl purge method PURGE
.if !'po4a'hide'       http_access deny purge !localhost
.if !'po4a'hide' .fi
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-n"
Proxy Negotiate(Kerberos) authentication.
.if !'po4a'hide' .nf
Use kinit username@DOMAIN first to get initial TGS.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-N"
WWW Negotiate(Kerberos) authentication.
.if !'po4a'hide' .nf
Use kinit username@DOMAIN first to get initial TGS.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-p port"
Port number of cache.  Default is 3128.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-P file"
Request body. Using the named file as data.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-r"
Force cache to reload URL.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-s"
Silent.  Do not print data to stdout.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-t count"
Trace
.I count
HTTP relay or proxy hops
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-T timeout"
Timeout value (seconds) for read/write operations.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-u user"
Proxy authentication username
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-U user"
WWW authentication username
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-v"
Verbose. Print outgoing message to stderr.
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-V version"
HTTP Version. Use '\-' for HTTP/0.9 omitted case
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-w password"
Proxy authentication password
.
.if !'po4a'hide' .TP
.if !'po4a'hide' .B "\-W password"
WWW authentication password
.
.SH AUTHOR
This program and manual was written by
.if !'po4a'hide' .I Amos Jeffries <amosjeffries@squid-cache.org>
.PP
Based on original code derived from Harvest and further developed by
numerous individuals from the internet community.
.
.SH COPYRIGHT
Distributed under the GNU General Public License (GNU GPL) version 2 or later (GPLv2+).
.
.SH QUESTIONS
Questions on the usage of this program can be sent to the
.I Squid Users mailing list
.if !'po4a'hide' <squid-users@squid-cache.org>
.
.SH REPORTING BUGS
See http://wiki.squid-cache.org/SquidFaq/BugReporting for details of what you need to include with your bug report.
.PP
Report bugs or bug fixes using http://bugs.squid-cache.org/
.PP
Report serious security bugs to
.I Squid Bugs <squid-bugs@squid-cache.org>
.PP
Report ideas for new improvements to the
.I Squid Developers mailing list
.if !'po4a'hide' <squid-dev@squid-cache.org>
.
.SH SEE ALSO
.if !'po4a'hide' .BR squid "(8), "
.if !'po4a'hide' .BR cachemgr.cgi "(8)"


ChangeLog
Changes to squid-3.5.0.2 (31 Oct 2014):

	- Fix FTP socket opening during reconfigure
	- ... and all changes from 3.4.9
	- ... and some build errors in rarely used code
	- ... and several documentation updates

Changes to squid-3.5.0.1 (17 Oct 2014):

	- Port from 2.7: redirector and logging urlgroup feature
	- Bug 4093: source-maintenance.sh bad perl -i option
	- Bug 3608: per-service name for workers UDS sockets
	- Bug 2554: 32-bit wrap in AUFS counters
	- Bug 1961 pt1: URL handling redesign
	- Bug 1202 pt1: documentation for refresh_pattern algorithms
	- Update Squid boilerplate copyright/license
	- Update the http(s)_port directives protocol= parameter
	- Update forward_max_tries to permit 25 server paths
	- Update Kerberos library detection and build options
	- Support ACLs on ftp_epsv directive
	- Support >32KB objects in cache_dir rock storage
	- Support client connection annotation by helpers via clt_conn_tag=TAG
	- Support native FTP Relay
	- Support libgnugss Kerberos library
	- Support libecap v1.0
	- Support SSL Peek and Splice feature
	- Support receiving PROXY protocol version 1 and 2
	- Replace --enable-ssl build option with --with-openssl
	- Enable -n service name command line option for all Squid builds
	- Enable ICAP client by default
	- Fix configuration file parsing bugs, related to quoted strings
	- Fix Windows MinGW build errors
	- Fix multiple TCP outgoing TOS/DiffServ bugs
	- Fix Cygwin /etc/resolv.conf parsing
	- Fix crash when sending %ssl::cert_subject to external ACL w/o certificate
	- Fix crash reading malformed config files
	- Send selected SSL version and cipher to the certificate validation helper
	- Validate server certificates without bumping
	- Add zero-copy string buffer support
	- Add automated squid.conf parser testing with squid -k parse
	- Add adaptation_service ACL
	- Add logformat code %tS to log transaction start time
	- Add logformat code %>rd to log client URL domain name
	- Add key_extras to proxy authentication
	- Add url_rewrite_extras and store_id_extras directives
	- Add send_hit and store_miss directives
	- Add collapsed_forwarding directive
	- Add sslproxy_cert_sign_hash directive
	- Add SMP SSL session cache
	- Add cache_peer standby connections
	- Add helper ext_delayer_acl
	- Add TCP_TUNNEL log code for CONNECT tunnels which are not SSL-bumped
	- Add BUILDCXX and BUILDCXXFLAGS configure options for cross-compile
	- Remove COSS storage in favour of Rock storage
	- Remove dnsserver and external DNS helper API in favour of mDNS
	- Remove broken mallinfo() accounting and memory tracing
	- Remove hierarchy_stoplist in favour of always_direct
	- Deprecate tag ACL type in favour of note ACL type
	- Deprecate urlgroup feature in favour of note ACL type
	- HTTP/1.1: method names are case-sensitive
	- HTTP/1.1: register new headers from RFC 723x
	- squidclient: polish and update help display
	- squidclient: support TLS with GnuTLS 3.1.5+
	- squidclient: support verbosity levels
	- squidclient: --ping mode module support
	- url_fake_rewrite: support concurrency
	- storeid_file_rewrite: support concurrency
	- digest_file_auth: support concurrency
	- digest_edirectory_auth: support concurrency
	- digest_ldap_auth: support concurrency
	- ... and many error page translation updates
	- ... and much code cleanup and polishing

Changes to squid-3.4.9 (31 Oct 2014):

	- Regression fix: ext_kerberos_ldap_group_acl typo in 3.4.7 update
	- Bug 4102: sslbump cert contains only a dot character in key usage extension
	- Bug 4093: source-maintenance.sh errors and warnings due to wrong tools/options
	- Bug 4088: memory leak in external_acl_type helper with cache=0 or ttl=0
	- Bug 4024: Bad host/IP ::1 when using IPv4-only environment
	- Bug 3803: ident leaks memory on failure
	- kerberos_ldap_group/cert_tool: Remove ksh dependency
	- ... and some automated code style updates
	- ... and some documentation updates

Changes to squid-3.4.8 (15 Sep 2014):

	- Fix off by one in SNMP subsystem
	- pinger: Fix various ICMP handling issues

Changes to squid-3.4.7 (28 Aug 2014):

	- Regression Fix: Kerberos LDAP authorizing groups with principle subdomain
	- Bug 4080: worker hangs when client identd is not responding
	- Bug 3966: Add KeyEncipherment when ssl-bump substitues RSA for EC
	- HTTP/1.1: Ignore Range headers with unidentifiable byte-range values
	- SSL-bump: Use v3 for fake certificate if we add _any_ certificate extension
	- Enable compile-time override for MAXTCPLISTENPORTS
	- ntlm_sspi_auth: Fix various build errors
	- negotiate_wrapper: Fix build issues with non-portable vfork()
	- negotiate_sspi_auth: Portability fixes for MinGW
	- ext_lm_group_acl: Portability fixes for MinGW
	- ... and several minor memory leaks

Changes to squid-3.4.6 (25 Jun 2014):

	- Regression: segmentation fault logging with %tg format specifier
	- Bug 4065: round-robin neighbor selection with unequal weights
	- Bug 4056: assertion MemPools[type] from netdbExchangeStart()
	- Bug 4050: segmentation fault in CommSelectEngine::checkEvents on helper response
	- Fix segmentation fault setting up server SSL connnection
	- Fix hanging Non-HTTPS connections on SSL-bump enabled port
	- Fix Cache Manager actions listed more than once
	- ... and many minor memory leaks
	- ... and several portability build issues
	- ... and some documentation updates

Changes to squid-3.4.5 (02 May 2014):

	- Regression Bug 4051: inverted test on CONNECT payload existence
	- Regression Fix: order dependency between cache_dir and maximum_object_size
	- Fix logformat %note display
	- Resolve 'dying from an unhandled exception: c'

Changes to squid-3.4.4.2 (23 Apr 2014):

	- version bump for packaging re-build with altered toolchain

Changes to squid-3.4.4.1 (23 Apr 2014):

	- Regression Bug 4019: Cache digest exchange segmentation fault
	- Regression Bug 3982: EUI logging and helpers show blank MAC address
	- Bug 4047: Support Android builds
	- Bug 4043: Remove XMALLOC_TRACE and references to sbrk(2)

tools/squidclient/squidclient.cc
/*
 * Copyright (C) 1996-2014 The Squid Software Foundation and contributors
 *
 * Squid software is distributed under GPLv2+ license and includes
 * contributions from numerous individuals and organizations.
 * Please see the COPYING and CONTRIBUTORS files for details.
 */

#include "squid.h"
#include "base64.h"
#include "ip/Address.h"
#include "ip/tools.h"
#include "rfc1123.h"
#include "tools/squidclient/gssapi_support.h"
#include "tools/squidclient/Parameters.h"
#include "tools/squidclient/Ping.h"
#include "tools/squidclient/Transport.h"

#if _SQUID_WINDOWS_
/** \cond AUTODOCS-IGNORE */
using namespace Squid;
/** \endcond */
#endif

#include <cerrno>
#include <csignal>
#include <cstring>
#include <iostream>
#if _SQUID_WINDOWS_
#include <io.h>
#endif
#if HAVE_SYS_SOCKET_H
#include <sys/socket.h>
#endif
#if HAVE_UNISTD_H
#include <unistd.h>
#endif
#if HAVE_NETDB_H
#include <netdb.h>
#endif
#if HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#if HAVE_FCNTL_H
#include <fcntl.h>
#endif
#if HAVE_NETINET_IN_H
#include <netinet/in.h>
#endif
#if HAVE_GETOPT_H
#include <getopt.h>
#endif

#ifndef BUFSIZ
#define BUFSIZ		8192
#endif
#ifndef MESSAGELEN
#define MESSAGELEN	65536
#endif
#ifndef HEADERLEN
#define HEADERLEN	65536
#endif

/* Local functions */
static void usage(const char *progname);

void pipe_handler(int sig);
static void set_our_signal(void);

Parameters scParams;

static int put_fd;
static char *put_file = NULL;

static struct stat sb;
int total_bytes = 0;

#if _SQUID_AIX_
/* Bug 3854: AIX 6.1 tries to link in this fde.h global symbol
 * despite squidclient not using any of the fd_* code.
 */
fde *fde::Table = NULL;
#endif

#if _SQUID_WINDOWS_
void
Win32SockCleanup(void)
{
    WSACleanup();
    return;
}
#endif

static void
usage(const char *progname)
{
    std::cerr << "Version: " << VERSION << std::endl
              << "Usage: " << progname << " [Basic Options] [HTTP Options]" << std::endl
              << std::endl;
    std::cerr
        << "    -s | --quiet    Silent.  Do not print response message to stdout." << std::endl
        << "    -v | --verbose  Verbose debugging. Repeat (-vv) to increase output level." << std::endl
        << "                    Levels:" << std::endl
        << "                      1 - Print outgoing request message to stderr." << std::endl
        << "                      2 - Print action trace to stderr." << std::endl
        << "    --help          Display this help text." << std::endl
        << std::endl;
    Transport::Config.usage();
    Ping::Config.usage();
    std::cerr
        << "HTTP Options:" << std::endl
        << "    -a           Do NOT include Accept: header." << std::endl
        << "    -A           User-Agent: header. Use \"\" to omit." << std::endl
        << "    -H 'string'  Extra headers to send. Use '\\n' for new lines." << std::endl
        << "    -i IMS       If-Modified-Since time (in Epoch seconds)." << std::endl
        << "    -j hosthdr   Host header content" << std::endl
        << "    -k           Keep the connection active. Default is to do only one request then close." << std::endl
        << "    -m method    Request method, default is GET." << std::endl
#if HAVE_GSSAPI
        << "    -n           Proxy Negotiate(Kerberos) authentication" << std::endl
        << "    -N           WWW Negotiate(Kerberos) authentication" << std::endl
#endif
        << "    -P file      Send content from the named file as request payload" << std::endl
        << "    -r           Force cache to reload URL" << std::endl
        << "    -t count     Trace count cache-hops" << std::endl
        << "    -u user      Proxy authentication username" << std::endl
        << "    -U user      WWW authentication username" << std::endl
        << "    -V version   HTTP Version. Use '-' for HTTP/0.9 omitted case" << std::endl
        << "    -w password  Proxy authentication password" << std::endl
        << "    -W password  WWW authentication password" << std::endl
        ;
    exit(1);
}

int
main(int argc, char *argv[])
{
    int len, bytesWritten;
    bool to_stdout, reload;
    int keep_alive = 0;
    int opt_noaccept = 0;
#if HAVE_GSSAPI
    int www_neg = 0, proxy_neg = 0;
#endif
    char url[BUFSIZ], msg[MESSAGELEN], buf[BUFSIZ];
    char extra_hdrs[HEADERLEN];
    const char *method = "GET";
    extern char *optarg;
    time_t ims = 0;
    int max_forwards = -1;

    const char *proxy_user = NULL;
    const char *proxy_password = NULL;
    const char *www_user = NULL;
    const char *www_password = NULL;
    const char *host = NULL;
    const char *version = "1.0";
    const char *useragent = NULL;

    /* set the defaults */
    extra_hdrs[0] = '\0';
    to_stdout = true;
    reload = false;

    Ip::ProbeTransport(); // determine IPv4 or IPv6 capabilities before parsing.
    if (argc < 2 || argv[argc-1][0] == '-') {
        usage(argv[0]);		/* need URL */
    } else if (argc >= 2) {
        strncpy(url, argv[argc - 1], BUFSIZ);
        url[BUFSIZ - 1] = '\0';

        int optIndex = 0;
        const char *shortOpStr = "aA:h:j:V:l:P:i:km:nNp:rsvt:H:T:u:U:w:W:?";

        // options for controlling squidclient
        static struct option basicOptions[] = {
            /* These are the generic options for squidclient itself */
            {"help",    no_argument, 0, '?'},
            {"verbose", no_argument, 0, 'v'},
            {"quiet",   no_argument, 0, 's'},
            {"host",    required_argument, 0, 'h'},
            {"local",   required_argument, 0, 'l'},
            {"port",    required_argument, 0, 'p'},
            {"ping",    no_argument, 0, '\1'},
            {"https",   no_argument, 0, '\3'},
            {0, 0, 0, 0}
        };

        int c;
        while ((c = getopt_long(argc, argv, shortOpStr, basicOptions, &optIndex)) != -1) {

            // modules parse their own specific options
            switch (c) {
            case '\1':
                to_stdout = 0;
                Ping::Config.parseCommandOpts(argc, argv, c, optIndex);
                continue;

            case 'h':		/* remote host */
            case 'l':		/* local host */
            case 'p':		/* port number */
                // rewind and let the Transport::Config parser handle
                optind -= 2;

            case '\3': // request over a TLS connection
                Transport::Config.parseCommandOpts(argc, argv, c, optIndex);
                continue;

            default: // fall through to next switch
                break;
            }

            switch (c) {

            case '\0': // dummy value for end-of-options
                break;

            case 'a':
                opt_noaccept = 1;
                break;

            case 'A':
                useragent = optarg;
                break;

            case 'j':
                host = optarg;
                break;

            case 'V':
                version = optarg;
                break;

            case 's':		/* silent */
                to_stdout = false;
                break;

            case 'k':		/* backward compat */
                keep_alive = 1;
                break;

            case 'r':		/* reload */
                reload = true;
                break;

            case 'P':
                put_file = xstrdup(optarg);
                break;

            case 'i':		/* IMS */
                ims = (time_t) atoi(optarg);
                break;

            case 'm':
                method = xstrdup(optarg);
                break;

            case 't':
                method = xstrdup("TRACE");
                max_forwards = atoi(optarg);
                break;

            case 'H':
                if (strlen(optarg)) {
                    char *t;
                    strncpy(extra_hdrs, optarg, sizeof(extra_hdrs));
                    while ((t = strstr(extra_hdrs, "\\n")))
                        *t = '\r', *(t + 1) = '\n';
                }
                break;

            case 'T':
                Transport::Config.ioTimeout = atoi(optarg);
                break;

            case 'u':
                proxy_user = optarg;
                break;

            case 'w':
                proxy_password = optarg;
                break;

            case 'U':
                www_user = optarg;
                break;

            case 'W':
                www_password = optarg;
                break;

            case 'n':
#if HAVE_GSSAPI
                proxy_neg = 1;
#else
                std::cerr << "ERROR: Negotiate authentication not supported." << std::endl;
                usage(argv[0]);
#endif
                break;

            case 'N':
#if HAVE_GSSAPI
                www_neg = 1;
#else
                std::cerr << "ERROR: Negotiate authentication not supported." << std::endl;
                usage(argv[0]);
#endif
                break;

            case 'v':
                /* undocumented: may increase verb-level by giving more -v's */
                ++scParams.verbosityLevel;
                debugVerbose(2, "verbosity level set to " << scParams.verbosityLevel);
                break;

            case '?':		/* usage */

            default:
                usage(argv[0]);
                break;
            }
        }
    }
#if _SQUID_WINDOWS_
    {
        WSADATA wsaData;
        WSAStartup(2, &wsaData);
        atexit(Win32SockCleanup);
    }
#endif
    /* Build the HTTP request */
    if (strncmp(url, "mgr:", 4) == 0) {
        char *t = xstrdup(url + 4);
        const char *at = NULL;
        if (!strrchr(t, '@')) { // ignore any -w password if @ is explicit already.
            at = proxy_password;
        }
        // embed the -w proxy password into old-style cachemgr URLs
        if (at)
            snprintf(url, BUFSIZ, "cache_object://%s/%s@%s", Transport::Config.hostname, t, at);
        else
            snprintf(url, BUFSIZ, "cache_object://%s/%s", Transport::Config.hostname, t);
        xfree(t);
    }
    if (put_file) {
        put_fd = open(put_file, O_RDONLY);
        set_our_signal();

        if (put_fd < 0) {
            std::cerr << "ERROR: can't open file (" << xstrerror() << ")" << std::endl;
            exit(-1);
        }
#if _SQUID_WINDOWS_
        setmode(put_fd, O_BINARY);
#endif

        if (fstat(put_fd, &sb) < 0) {
            std::cerr << "ERROR: can't identify length of file (" << xstrerror() << ")" << std::endl;
        }
    }

    if (!host) {
        char *newhost = strstr(url, "://");
        if (newhost) {
            char *t;
            newhost += 3;
            newhost = xstrdup(newhost);
            t = newhost + strcspn(newhost, "@/?");
            if (*t == '@') {
                newhost = t + 1;
                t = newhost + strcspn(newhost, "@/?");
            }
            *t = '\0';
            host = newhost;
        }
    }

    if (version[0] == '-' || !version[0]) {
        /* HTTP/0.9, no headers, no version */
        snprintf(msg, BUFSIZ, "%s %s\r\n", method, url);
    } else {
        if (!xisdigit(version[0])) // not HTTP/n.n
            snprintf(msg, BUFSIZ, "%s %s %s\r\n", method, url, version);
        else
            snprintf(msg, BUFSIZ, "%s %s HTTP/%s\r\n", method, url, version);

        if (host) {
            snprintf(buf, BUFSIZ, "Host: %s\r\n", host);
            strcat(msg,buf);
        }

        if (useragent == NULL) {
            snprintf(buf, BUFSIZ, "User-Agent: squidclient/%s\r\n", VERSION);
            strcat(msg,buf);
        } else if (useragent[0] != '\0') {
            snprintf(buf, BUFSIZ, "User-Agent: %s\r\n", useragent);
            strcat(msg,buf);
        }

        if (reload) {
            snprintf(buf, BUFSIZ, "Cache-Control: no-cache\r\n");
            strcat(msg, buf);
        }
        if (put_fd > 0) {
            snprintf(buf, BUFSIZ, "Content-length: %" PRId64 "\r\n", (int64_t) sb.st_size);
            strcat(msg, buf);
        }
        if (opt_noaccept == 0) {
            snprintf(buf, BUFSIZ, "Accept: */*\r\n");
            strcat(msg, buf);
        }
        if (ims) {
            snprintf(buf, BUFSIZ, "If-Modified-Since: %s\r\n", mkrfc1123(ims));
            strcat(msg, buf);
        }
        if (max_forwards > -1) {
            snprintf(buf, BUFSIZ, "Max-Forwards: %d\r\n", max_forwards);
            strcat(msg, buf);
        }
        if (proxy_user) {
            const char *user = proxy_user;
            const char *password = proxy_password;
#if HAVE_GETPASS
            if (!password)
                password = getpass("Proxy password: ");
#endif
            if (!password) {
                std::cerr << "ERROR: Proxy password missing" << std::endl;
                exit(1);
            }
            snprintf(buf, BUFSIZ, "%s:%s", user, password);
            snprintf(buf, BUFSIZ, "Proxy-Authorization: Basic %s\r\n", old_base64_encode(buf));
            strcat(msg, buf);
        }
        if (www_user) {
            const char *user = www_user;
            const char *password = www_password;
#if HAVE_GETPASS
            if (!password)
                password = getpass("WWW password: ");
#endif
            if (!password) {
                std::cerr << "ERROR: WWW password missing" << std::endl;
                exit(1);
            }
            snprintf(buf, BUFSIZ, "%s:%s", user, password);
            snprintf(buf, BUFSIZ, "Authorization: Basic %s\r\n", old_base64_encode(buf));
            strcat(msg, buf);
        }
#if HAVE_GSSAPI
        if (www_neg) {
            if (host) {
                snprintf(buf, BUFSIZ, "Authorization: Negotiate %s\r\n", GSSAPI_token(host));
                strcat(msg, buf);
            } else
                std::cerr << "ERROR: server host missing" << std::endl;
        }
        if (proxy_neg) {
            if (Transport::Config.hostname) {
                snprintf(buf, BUFSIZ, "Proxy-Authorization: Negotiate %s\r\n", GSSAPI_token(Transport::Config.hostname));
                strcat(msg, buf);
            } else
                std::cerr << "ERROR: proxy server host missing" << std::endl;
        }
#endif

        /* HTTP/1.0 may need keep-alive explicitly */
        if (strcmp(version, "1.0") == 0 && keep_alive)
            strcat(msg, "Connection: keep-alive\r\n");

        /* HTTP/1.1 may need close explicitly */
        if (!keep_alive)
            strcat(msg, "Connection: close\r\n");

        strcat(msg, extra_hdrs);
        strcat(msg, "\r\n");
    }

    debugVerbose(1, "Request:" << std::endl << msg << std::endl << ".");

    uint32_t loops = Ping::Init();

    for (uint32_t i = 0; loops == 0 || i < loops; ++i) {
        size_t fsize = 0;

        if (!Transport::Connect())
            continue;

        /* Send the HTTP request */
        debugVerbose(2, "Sending HTTP request ... ");
        bytesWritten = Transport::Write(msg, strlen(msg));

        if (bytesWritten < 0) {
            std::cerr << "ERROR: write" << std::endl;
            exit(1);
        } else if ((unsigned) bytesWritten != strlen(msg)) {
            std::cerr << "ERROR: Cannot send request?: " << std::endl << msg << std::endl;
            exit(1);
        }
        debugVerbose(2, "done.");

        if (put_file) {
            debugVerbose(1, "Sending HTTP request payload ...");
            int x;
            lseek(put_fd, 0, SEEK_SET);
            while ((x = read(put_fd, buf, sizeof(buf))) > 0) {

                x = Transport::Write(buf, x);

                total_bytes += x;

                if (x <= 0)
                    break;
            }

            if (x != 0)
                std::cerr << "ERROR: Cannot send file." << std::endl;
            else
                debugVerbose(1, "done.");
        }
        /* Read the data */

#if _SQUID_WINDOWS_
        setmode(1, O_BINARY);
#endif

        while ((len = Transport::Read(buf, sizeof(buf))) > 0) {
            fsize += len;

            if (to_stdout && fwrite(buf, len, 1, stdout) != 1)
                std::cerr << "ERROR: writing to stdout: " << xstrerror() << std::endl;
        }

#if USE_GNUTLS
        if (Transport::Config.tlsEnabled) {
            if (len == 0) {
                std::cerr << "- Peer has closed the TLS connection" << std::endl;
            } else if (!gnutls_error_is_fatal(len)) {
                std::cerr << "WARNING: " << gnutls_strerror(len) << std::endl;
            } else {
                std::cerr << "ERROR: " << gnutls_strerror(len) << std::endl;
            }
        }
#endif

#if _SQUID_WINDOWS_
        setmode(1, O_TEXT);
#endif

        Transport::CloseConnection();

        if (Ping::LoopDone(i))
            break;

        Ping::TimerStop(fsize);
    }

    Ping::DisplayStats();
    Transport::ShutdownTls();
    return 0;
}

void
pipe_handler(int sig)
{
    std::cerr << "SIGPIPE received." << std::endl;
}

static void
set_our_signal(void)
{
#if HAVE_SIGACTION
    struct sigaction sa;
    sa.sa_handler = pipe_handler;
    sa.sa_flags = SA_RESTART;
    sigemptyset(&sa.sa_mask);

    if (sigaction(SIGPIPE, &sa, NULL) < 0) {
        std::cerr << "ERROR: Cannot set PIPE signal." << std::endl;
        exit(-1);
    }
#else
    signal(SIGPIPE, pipe_handler);
#endif
}
